### PR TITLE
EphemeralBuffers for Windows and Unix, where the memory won't be page…

### PR DESCRIPTION
**Pull request by Drawaes**
Why ? 

Well storing perm secrets should be done on things like TPM's, however for items like session keys it is normally okay to store them unencrypted in memory. However perfect forward secrecy for instance can be badly compromised if the memory is paged out to disk, as it can hang around for a long time and is easier to attack than crossing into a processes memory space.

### CODE REVIEW

**Comment by Drawaes** [comment]
The buffers are zero'ed on return, and on dispose of the entire pool. VirtualLock on windows is used and MLock on unix. For the Interop I used (although needed to add to ) files from the PAL layers in Corefx, so it should be  compliant generally there.

**Comment by Drawaes** [comment]
@davidfowl these are what we talked about for in memory secrets

**Comment by KrzysztofCwalina** [comment]
System.Buffers.Experimental is a pure portable library. I would not want interop code in it. I will let @davidfowl review the rest of the PR, but if he does want to add it to corfxlab, it should be in a different package, not in System.Buffers.Experimental.

File: src/System.Buffers.Experimental/System/Buffers/EphemeralBufferWindows.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            if (!buffer2.Rented)
+            {
+                return;
+            }
+            buffer2.Rented = false;
+            _emptyData.CopyTo(buffer.Span);
+            _buffers.Enqueue(buffer2);
+        }
+
+        sealed class SecureMemory : OwnedMemory<byte>
```
Comment (benaadams) [inline src/System.Buffers.Experimental/System/Buffers/EphemeralBufferWindows.cs]: `EphemeralMemory` 😛

Comment (Drawaes) [inline src/System.Buffers.Experimental/System/Buffers/EphemeralBufferWindows.cs]: Whoops, missed that in the rename will change.

**Comment by Drawaes** [comment]
Sure it can be moved, just seemed like it fit will all the other buffers, where would it go... ?

**Comment by KrzysztofCwalina** [comment]
I would not mind a separate package/dll, e.g. System.Buffers.Ephemeral. Having said that, don't do too much work till @davidfowl confirms that he wants this in corfxlab (unless you are sure).

**Comment by Drawaes** [comment]
I am not sure :) I am using it in my library currently and will continue to do so until/if it becomes part of the lab.

**Comment by davidfowl** [comment]
Seems pretty specialized to go into System.Buffers... It should stay in  the SSL lib for now.

**Comment by Drawaes** [comment]
Fair enough

**Pull request closed without merging**